fix(flow): Prevent initial_state BaseModel from being coerced to dict

diego-coder opened the pull request:
Fixes #3147

This PR resolves a critical bug where passing a Pydantic `BaseModel` as `initial_state` to a `Flow` would cause it to be silently coerced into a dictionary. This led to attribute collisions and runtime errors if the `BaseModel` contained fields with names matching dictionary methods (e.g., `items`, `keys`, `values`).

The fix ensures that the `BaseModel` type is preserved throughout the flow's lifecycle by using Pydantic's native copy methods and correcting the `Flow` constructor.

#### Key Changes & Fixes

1.  **Corrected `Flow` Constructor:** The `__init__` method in `src/crewai/flow/flow.py` was modified to properly accept the `initial_state` parameter. Previously, this parameter was being ignored and incorrectly handled, which was the root cause of the state being initialized as a dictionary.

2.  **Preserved `BaseModel` on Creation:** The `_create_initial_state` method was updated to use `model.model_copy()` (Pydantic v2) or `model.copy()` (Pydantic v1) instead of converting the model to a dictionary. This ensures the state is a `BaseModel` instance from the very beginning.

3.  **Safe `BaseModel` Updates:** The `_initialize_state` method was updated to use `model.copy(update=...)` for applying inputs. This is a much safer and more direct way to update Pydantic models than the previous dictionary-merging approach.

#### Changes to Tests

A new, comprehensive test suite was created in `tests/test_flow_initial_state_fix.py` to validate the fix and prevent regressions:

1.  **Reproduces the Bug:** The tests confirm the original bug by using state models with colliding attribute names like `items`, `keys`, `values`, and `get`.
2.  **Validates the Fix:** The assertions now pass, proving that `flow.state` remains a `BaseModel` instance and that attributes can be accessed correctly.
3.  **Corrected Test Models:** All Pydantic models used for testing were updated to inherit from `FlowState`. This was a necessary change because the fix correctly enforces the requirement that all state models must have an `id` field, a detail the initial automated PR's tests missed.

#### Comparison to Previous Automated PR
The initial automated PR (#3148) correctly identified the need for `model_copy()` but failed because it missed two critical issues that this PR addresses:
1.  It did not fix the `Flow` constructor, which was the primary source of the bug.
2.  Its tests were incomplete and failed because they did not account for the mandatory `id` field in state models.

#### Code Quality & Linting
All modified files (`src/crewai/flow/flow.py` and `tests/test_flow_initial_state_fix.py`) have been formatted with `ruff format` and checked with `ruff check --fix` to ensure they adhere to the project's coding standards.

lucasgomide commented on tests/test_flow_initial_state_fix.py:
```diff
@@ -0,0 +1,99 @@
+"""Test Flow initial_state BaseModel dict coercion fix for issue #3147"""
```
Can you move the tests to `flow_test.py` I really dont think we should create another file for those tests

Also, would you mind removing unnecessary comments?

diego-coder commented:
Sounds good @lucasgomide, changed it:
    Restored the docstring for the ensure_state_type function.
    Moved the new tests into flow_test.py and removed the separate test file.
    Reverted the logic in the _initialize_state method back to the model_dump approach.
The PR should be ready for another look.
